added support for case_weights

p-schaefer opened the pull request:
I noticed there wasn't much progress on integrating case_weights into lightgbm through bonsai so I've taken a crack at adding it. This is in reference to [65](https://github.com/tidymodels/bonsai/issues/65).

simonpcouch commented on tests/testthat/test-lightgbm.R:
```diff
@@ -700,3 +700,109 @@ test_that("multi_predict() predicts classes if 'type' not given ", {
+                             , seed = 1234L
+                             , deterministic = "true"
+                             , force_col_wise = "true"
+                             ) %>%
```
In addition to the style edits mentioned above, it'd be great if we could pare back the arguments set here as much as possible.

p-schaefer replied:
I pared the example down as suggested.

simonpcouch replied:
Thanks!

simonpcouch commented on R/lightgbm.R:
```diff
@@ -35,7 +36,7 @@
 #' @keywords internal
 #' @export
-train_lightgbm <- function(x, y, max_depth = -1, num_iterations = 100, learning_rate = 0.1,
+train_lightgbm <- function(x, y,weights=NULL, max_depth = -1, num_iterations = 100, learning_rate = 0.1,
```
https://style.tidyverse.org/syntax.html

p-schaefer replied:
I think I've corrected these

simonpcouch reviewed (commented):
A strong start! Some pointers to move this closer:

p-schaefer commented on R/lightgbm.R:
```diff
@@ -224,18 +225,20 @@ process_data <- function(args, x, y, validation, missing_validation,
   if (!is.null(val_index)) {
     args$main$valids <-
       list(validation =
-          lightgbm::lgb.Dataset(
```
I think I've corrected these? Let me know if thats not the case.

simonpcouch replied:
Almost there. If you look in this diff, there are several spaces before the function call and arguments that didn't used to be there.

p-schaefer commented:
Thanks for the review @simonpcouch. Note, I am also working on a method that allows some ability to customize validation data beyond just specifying a proportion of training data. I find this quite necessary for my work (since my data are not entirely independent - hence needing case weights as well), but I don't know if its robust enough to include in Bonsai proper.

simonpcouch commented on R/lightgbm.R:
```diff
@@ -281,11 +284,11 @@ sort_args <- function(args) {
 # this function ensures that multiclass classification predictions are always
 # returned as a [num_observations, num_classes] matrix, regardless of lightgbm version
 reshape_lightgbm_multiclass_preds <- function(preds, num_rows) {
-    n_preds_per_case <- length(preds) / num_rows
```
More space deletions here that are unrelated to the PR.

simonpcouch commented on R/lightgbm.R:
```diff
@@ -324,9 +327,9 @@ predict_lightgbm_classification_class <- function(object, new_data, ...) {
-      p <- stats::predict(object$fit, prepare_df_lgbm(new_data), type = "raw", ...)
+    p <- stats::predict(object$fit, prepare_df_lgbm(new_data), type = "raw", ...)
   } else {
-      p <- stats::predict(object$fit, prepare_df_lgbm(new_data), rawscore = TRUE, ...)
```
Here as well.

simonpcouch commented on tests/testthat/_snaps/lightgbm.new.md:
This file, `lightgbm.new.md`, is an intermediate file from testthat. To merge its changes into the existing `test-lightgbm.R`, use `testthat::snapshot_accept()`.

simonpcouch commented:
Thanks again for your work here. Given the lack of movement on this PR, I'm going to go ahead and close. :)

simonpcouch commented:
That'd be wonderful, @joranE! Conceptually, the implementation here looks solid—following the current review comments would put this PR in a good spot.

p-schaefer commented:
Sorry, this fell off my radar. I can finish addressing the comments next week, if that works for you.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Simon P. Couch ***@***.***>
Sent: Monday, April 8, 2024 9:31:10 AM
To: tidymodels/bonsai ***@***.***>
Cc: Patrick Schaefer ***@***.***>; Author ***@***.***>
Subject: Re: [tidymodels/bonsai] added support for case_weights (PR #72)


That'd be wonderful, @joranE<https://github.com/joranE>! Conceptually, the implementation here looks solid—following the current review comments would put this PR in a good spot.

—
Reply to this email directly, view it on GitHub<https://github.com/tidymodels/bonsai/pull/72#issuecomment-2042767586>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACYA7SXU5VRFVBVDIZ322P3Y4KLZ5AVCNFSM6AAAAABAUFJUD6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBSG43DONJYGY>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

joranE commented:
@p-schaefer Sounds good, I'll let you tackle it! If you find yourself pulled away from it again for any reason, just let me know.

p-schaefer commented:
I've fixed the spacing issues, and dealt with the intermediate file from testthat. Should I submit a new PR?

simonpcouch commented:
You're welcome to push to the same branch! I'll reopen this one.

p-schaefer commented:
Sorry, forgot to rerun the tests and checks. Looks good to me now, let me know if there is anything else.

joranE commented:
Thanks for closing the loop on this @p-schaefer! Now I just need to figure out if there's an easy way to tweak this interface to allow for the use of the `linear_tree` parameter for piecewise linear trees, which "naturally" needs to be passed to `lgb.Dataset()` as well, instead of `lgb.train()`. 🙄

joranE commented:
I added a new issue [here](https://github.com/tidymodels/bonsai/issues/77).

simonpcouch commented on tests/testthat/test-lightgbm.R:
```diff
@@ -700,3 +700,74 @@ test_that("multi_predict() predicts classes if 'type' not given ", {
+  data(agaricus.train, package = "lightgbm")
+  dtrain <- lgb.Dataset(agaricus.train$data, label = agaricus.train$label, weight = weights1)
+  train <- data.frame(as.matrix(agaricus.train$data))
+  train$label<-agaricus.train$label
```
Could you do a pass through these tests enforcing [tidy style](https://style.tidyverse.org/syntax.html)?
```suggestion
  train$label <- agaricus.train$label
```

Namely, spaces before and after infix operators.

p-schaefer replied:
I hope I got them all this time.

simonpcouch commented:
Test failures are unrelated, feel free to ignore. :)

p-schaefer commented:
@simonpcouch Was there anything else you need from me?

simonpcouch commented:
Nice work, @p-schaefer! :)
